### PR TITLE
type: refactor (external) code management

### PR DESCRIPTION
**Pull request by TeXitoi**

### CODE REVIEW

File: source/type/code_container.h [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+struct Network;
+struct Company;
+struct Line;
+struct Route;
+struct VehicleJourney;
+struct StopPoint;
+struct Calendar;
+
+/**
+ * Code container
```
Comment (antoine-de) [inline source/type/code_container.h]: It might be nice to add a bit a comments there

File: source/type/code_container.h [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        auto& obj_m = at_key<T>(obj_map);
+        obj_m[obj][key].push_back(val);
+
+        auto& code_m = at_key<T>(code_map);
+        code_m[{key, val}].push_back(obj);
+    }
+
+    template<class Archive>
+    void serialize(Archive& ar, const unsigned int) {
+        ar & obj_map;
```
Comment (antoine-de) [inline source/type/code_container.h]: you forgot code_map :boom:

File: source/tests/utils_test.h [hunk 3]
```diff
@@ -3,6 +3,10 @@
 #include "type/datetime.h"
 #include <boost/date_time/gregorian/gregorian.hpp>
 #include <boost/date_time/posix_time/posix_time.hpp>
+#include <set>
+#include <vector>
+#include <map>
```
Comment (xlqian) [inline source/tests/utils_test.h:8]: unused include?

File: source/jormungandr/tests/ptref_tests.py [hunk 4]
```diff
@@ -328,19 +328,29 @@ def test_vj_period_filter(self):
...
 
 
+@dataset(["main_ptref_test", "main_routing_test"])
+class TestPtRefRoutingAndPtrefCov(AbstractTestFixture):
+    def test_external_code(self):
+        """test the strange and ugly external code api"""
+        response = self.query("v1/lines?external_code=A&show_codes=true")
+        lines = get_not_null(response, 'lines')
+        assert len(lines) == 1
+        assert 'A' in [code['value'] for code in lines[0]['codes'] if code['type'] == 'external_code']
```
Comment (xlqian) [inline source/jormungandr/tests/ptref_tests.py:349]: use filter

File: source/jormungandr/tests/ptref_tests.py [hunk 5]
```diff
@@ -184,9 +182,11 @@ def test_route(self):
 
         routes = get_not_null(response, 'routes')
 
-        assert len(routes) == 2
+        assert len(routes) == 3
 
-        r = routes[0]
+        r = [r for r in routes if r['id'] == 'line:A:0']
```
Comment (xlqian) [inline source/jormungandr/tests/ptref_tests.py:187]: use filter 

``` python
r = filter(lambda r: r['id']=='line:A:0', routes)
```

File: source/type/code_container.h [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        PairByObj<Company>,
+        PairByObj<Line>,
+        PairByObj<Route>,
+        PairByObj<VehicleJourney>,
+        PairByObj<StopPoint>,
+        PairByObj<Calendar>
+        >;
+
+    template<typename T>
+    using ObjsByCodeMap = std::map<std::pair<std::string, std::string>, Objs<T>>;
```
Comment (xlqian) [inline source/type/code_container.h:113]: why using a pair of string instead of one concatenated string?

File: source/type/code_container.h [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+    template<typename T>
+    using RemoveConstPtr = typename std::remove_const<typename std::remove_pointer<T>::type>::type;
+
+    template <typename T, typename Sequence>
+    static auto at_key(Sequence& seq) -> decltype(boost::fusion::at_key<RemoveConstPtr<T>>(seq)) {
+        return boost::fusion::at_key<RemoveConstPtr<T>>(seq);
+    }
+
+    ObjMap obj_map;
```
Comment (xlqian) [inline source/type/code_container.h]: unused object? if so, the definition ObjMap is also not used?

File: source/tests/utils_test.h (hunk 3 above)
Comment (TeXitoi) [inline source/tests/utils_test.h:8]: nop, needed.

File: source/jormungandr/tests/ptref_tests.py (hunk 5 above)
Comment (TeXitoi) [inline source/jormungandr/tests/ptref_tests.py:187]: One char bigger... ;)

File: source/type/code_container.h (hunk 6 above)
Comment (TeXitoi) [inline source/type/code_container.h:113]: It doesn't work.

**Pull request merged**